File: sdk/python/trialmcp_client/provenance.py
"""Provenance sub-client for data lineage operations.

Wraps the four Provenance server tools: ``provenance_record``,
``provenance_query_forward``, ``provenance_query_backward``, and
``provenance_verify``.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from .config import ServerName
from .exceptions import raise_for_error
from .models import ProvenanceRecord, ProvenanceVerification

if TYPE_CHECKING:
    from .client import TrialMCPClient

logger = logging.getLogger("trialmcp.provenance")

_SERVER = ServerName.PROVENANCE


class ProvenanceClient:
    """Sub-client for the ``trialmcp-provenance`` MCP server.

    Provides typed, async wrappers around the four provenance tools.
    The provenance server maintains a directed acyclic graph (DAG) of
    data lineage events for clinical-trial data integrity.

    Parameters
    ----------
    parent:
        The :class:`TrialMCPClient` that owns this sub-client.
    """

    def __init__(self, parent: TrialMCPClient) -> None:
        self._client = parent

    # ------------------------------------------------------------------
    # provenance_record
    # ------------------------------------------------------------------

    async def record(
        self,
        *,
        source_id: str,
        source_type: str,
        action: str,
        actor_id: str,
        actor_role: str,
        tool_call: str,
        input_hash: str | None = None,
        output_hash: str | None = None,
        origin_server: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> ProvenanceRecord:
        """Record a new data access or transformation event.

        Parameters
        ----------
        source_id:
            Reference to the DataSource being accessed (UUID).
        source_type:
            Category of the data source (e.g. ``"fhir_resource"``).
        action:
            Operation performed (``"read"``, ``"transform"``, etc.).
        actor_id:
            Identifier of the actor performing the action.
        actor_role:
            Role of the actor from the 6-role model.
        tool_call:
            MCP tool invocation that triggered this event.
        input_hash:
            SHA-256 fingerprint of the input data.
        output_hash:
            SHA-256 fingerprint of the output data.
        origin_server:
            MCP server that owns the data source.
        description:
            Human-readable description of the event.
        metadata:
            Additional properties specific to the source type.

        Returns
        -------
        ProvenanceRecord:
            The newly created provenance record.
        """
        params: dict[str, Any] = {
            "source_id": source_id,
            "source_type": source_type,
            "action": action,
            "actor_id": actor_id,
            "actor_role": actor_role,
            "tool_call": tool_call,
        }
        if input_hash:
            params["input_hash"] = input_hash
        if output_hash:
            params["output_hash"] = output_hash
        if origin_server:
            params["origin_server"] = origin_server
        if description:
            params["description"] = description
        if metadata:
            params["metadata"] = metadata

        result = await self._client.call_tool(_SERVER, "provenance_record", params)
        raise_for_error(result)
        return ProvenanceRecord.from_dict(result)

    # ------------------------------------------------------------------
    # provenance_query_forward
    # ------------------------------------------------------------------

    async def query_forward(
        self,
        *,
        source_id: str,
        max_depth: int = 10,
        action_filter: str | None = None,
    ) -> list[ProvenanceRecord]:
        """Trace data lineage forward from a source (what was derived from it).

        Parameters
        ----------
        source_id:
            Starting data source ID.
        max_depth:
            Maximum DAG traversal depth.
        action_filter:
            Optional filter by action type.

        Returns
        -------
        list[ProvenanceRecord]:
            Ordered list of downstream provenance records.
        """
        params: dict[str, Any] = {
            "source_id": source_id,
            "max_depth": max_depth,
        }
        if action_filter:
            params["action_filter"] = action_filter

        result = await self._client.call_tool(_SERVER, "provenance_query_forward", params)
        raise_for_error(result)

        records_raw = result.get("records", [])
        return [ProvenanceRecord.from_dict(r) for r in records_raw]

    # ------------------------------------------------------------------
    # provenance_query_backward
    # ------------------------------------------------------------------

    async def query_backward(
        self,
        *,
        record_id: str,
        max_depth: int = 10,
        action_filter: str | None = None,
    ) -> list[ProvenanceRecord]:
        """Trace data lineage backward to its origins.

        Parameters
        ----------
        record_id:
            Starting provenance record ID.
        max_depth:
            Maximum DAG traversal depth.
        action_filter:
            Optional filter by action type.

        Returns
        -------
        list[ProvenanceRecord]:
            Ordered list of upstream provenance records.
        """
        params: dict[str, Any] = {
            "record_id": record_id,
            "max_depth": max_depth,
        }
        if action_filter:
            params["action_filter"] = action_filter

        result = await self._client.call_tool(_SERVER, "provenance_query_backward", params)
        raise_for_error(result)

        records_raw = result.get("records", [])
        return [ProvenanceRecord.from_dict(r) for r in records_raw]

    # ------------------------------------------------------------------
    # provenance_verify
    # ------------------------------------------------------------------

    async def verify(
        self,
        *,
        source_id: str | None = None,
        record_id: str | None = None,
        verify_hashes: bool = True,
    ) -> ProvenanceVerification:
        """Verify the integrity of provenance records.

        Checks SHA-256 fingerprints and DAG link consistency.

        Parameters
        ----------
        source_id:
            Verify all records for a specific data source.
        record_id:
            Verify a specific provenance record and its chain.
        verify_hashes:
            Whether to verify SHA-256 fingerprints.

        Returns
        -------
        ProvenanceVerification:
            Verification result with pass/fail details.
        """
        params: dict[str, Any] = {
            "verify_hashes": verify_hashes,
        }
        if source_id:
            params["source_id"] = source_id
        if record_id:
            params["record_id"] = record_id

        result = await self._client.call_tool(_SERVER, "provenance_verify", params)
        raise_for_error(result)

        return ProvenanceVerification(
            valid=result.get("valid", False),
            records_checked=result.get("records_checked", 0),
            broken_links=result.get("broken_links", []),
            error_message=result.get("error_message"),
        )

    # ------------------------------------------------------------------
    # Convenience helpers
    # ------------------------------------------------------------------

    async def record_read_access(
        self,
        *,
        source_id: str,
        source_type: str,
        actor_id: str,
        actor_role: str,
        tool_call: str,
        origin_server: str,
        output_hash: str | None = None,
        description: str | None = None,
    ) -> ProvenanceRecord:
        """Record a read-access event (common shorthand).

        Parameters
        ----------
        source_id:
            The data source being read.
        source_type:
            Category of the data source.
        actor_id:
            Actor performing the read.
        actor_role:
            Role of the actor.
        tool_call:
            MCP tool that performed the read.
        origin_server:
            Server that owns the data.
        output_hash:
            SHA-256 fingerprint of the read data.
        description:
            Human-readable description.

        Returns
        -------
        ProvenanceRecord:
            The recorded provenance event.
        """
        return await self.record(
            source_id=source_id,
            source_type=source_type,
            action="read",
            actor_id=actor_id,
            actor_role=actor_role,
            tool_call=tool_call,
            origin_server=origin_server,
            output_hash=output_hash,
            description=description,
        )

    async def full_lineage(
        self,
        *,
        source_id: str,
    ) -> dict[str, list[ProvenanceRecord]]:
        """Retrieve both forward and backward lineage for a data source.

        Parameters
        ----------
        source_id:
            The data source ID to trace.

        Returns
        -------
        dict:
            ``{"forward": [...], "backward": [...]}`` provenance chains.
        """
        import asyncio

        forward_task = self.query_forward(source_id=source_id)
        backward_task = self.query_backward(record_id=source_id)

        forward_records, backward_records = await asyncio.gather(forward_task, backward_task)

        return {
            "forward": forward_records,
            "backward": backward_records,
        }

    async def verify_source(self, source_id: str) -> ProvenanceVerification:
        """Verify all provenance records for a specific data source.

        Parameters
        ----------
        source_id:
            The data source ID.

        Returns
        -------
        ProvenanceVerification:
            Verification result.
        """
        return await self.verify(source_id=source_id)

File: sdk/typescript/src/provenance.ts
/**
 * Provenance client for the National MCP-PAI Oncology Trials network.
 *
 * Wraps the trialmcp-provenance MCP server tools:
 *   - provenance_record_access: Record a provenance event for data access
 *   - provenance_query_forward: Trace data lineage forward from a source
 *   - provenance_query_backward: Trace data lineage backward to origins
 *   - provenance_verify: Verify provenance chain integrity for a source
 *
 * Provenance tracking enables full data lineage across the oncology
 * trial network, supporting regulatory audit requirements (21 CFR Part 11,
 * GDPR Article 30) and W3C PROV-DM compliance.
 */

import { TrialMcpClient } from './client';
import { InvalidInputError } from './errors';
import type {
  ProvenanceRecord,
  ProvenanceRecordParams,
  ProvenanceQueryResult,
  ProvenanceVerifyResult,
  ActorRole,
} from './models';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const SERVER: 'trialmcp-provenance' = 'trialmcp-provenance';

/** Standard provenance actions aligned with W3C PROV-DM */
const PROVENANCE_ACTIONS = [
  'CREATE',
  'READ',
  'UPDATE',
  'DELETE',
  'TRANSFORM',
  'DERIVE',
  'TRANSFER',
  'VALIDATE',
  'APPROVE',
  'SIGN',
  'ANONYMIZE',
  'PSEUDONYMIZE',
] as const;

type ProvenanceAction = typeof PROVENANCE_ACTIONS[number];

/** Valid actor roles */
const VALID_ROLES: ActorRole[] = [
  'robot_agent',
  'trial_coordinator',
  'data_monitor',
  'auditor',
  'sponsor',
  'cro',
];

// ---------------------------------------------------------------------------
// Provenance Client
// ---------------------------------------------------------------------------

export class ProvenanceClient {
  private readonly client: TrialMcpClient;

  constructor(client: TrialMcpClient) {
    this.client = client;
  }

  /**
   * Record a provenance event for a data access or transformation.
   *
   * Creates an immutable provenance record linking an actor, action,
   * and data source. Supports input/output hashes for data integrity
   * tracking across transformations.
   *
   * @param params - Provenance record parameters
   * @returns The created provenance record with server-assigned ID
   */
  async record(params: ProvenanceRecordParams): Promise<ProvenanceRecord> {
    this.validateRecordParams(params);

    const toolParams: Record<string, unknown> = {
      source_id: params.source_id,
      action: params.action,
      actor_id: params.actor_id,
      actor_role: params.actor_role,
      tool_call: params.tool_call,
    };

    if (params.source_type) toolParams.source_type = params.source_type;
    if (params.input_hash) toolParams.input_hash = params.input_hash;
    if (params.output_hash) toolParams.output_hash = params.output_hash;
    if (params.origin_server) toolParams.origin_server = params.origin_server;
    if (params.description) toolParams.description = params.description;
    if (params.metadata) toolParams.metadata = params.metadata;

    const result = await this.client.callTool<ProvenanceRecord>(
      SERVER,
      'provenance_record_access',
      toolParams,
    );

    return result.data;
  }

  /**
   * Record a data read provenance event.
   * Convenience method for the most common provenance action.
   */
  async recordRead(
    sourceId: string,
    actorId: string,
    actorRole: ActorRole,
    toolCall: string,
    description?: string,
  ): Promise<ProvenanceRecord> {
    return this.record({
      source_id: sourceId,
      action: 'READ',
      actor_id: actorId,
      actor_role: actorRole,
      tool_call: toolCall,
      description,
    });
  }

  /**
   * Record a data transformation provenance event.
   * Captures input and output hashes for integrity verification.
   */
  async recordTransform(
    sourceId: string,
    actorId: string,
    actorRole: ActorRole,
    toolCall: string,
    inputHash: string,
    outputHash: string,
    description?: string,
  ): Promise<ProvenanceRecord> {
    return this.record({
      source_id: sourceId,
      action: 'TRANSFORM',
      actor_id: actorId,
      actor_role: actorRole,
      tool_call: toolCall,
      input_hash: inputHash,
      output_hash: outputHash,
      description,
    });
  }

  /**
   * Query provenance records forward from a source.
   *
   * Traces the lineage of data from a specific source forward in time,
   * showing how the data was used, transformed, or derived.
   *
   * @param sourceId - Source identifier to trace from
   * @param maxDepth - Maximum depth of forward traversal (default: 10)
   * @returns Forward provenance chain from the source
   */
  async queryForward(
    sourceId: string,
    maxDepth: number = 10,
  ): Promise<ProvenanceQueryResult> {
    this.validateSourceId(sourceId);
    this.validateDepth(maxDepth);

    const result = await this.client.callTool<ProvenanceQueryResult>(
      SERVER,
      'provenance_query_forward',
      { source_id: sourceId, max_depth: maxDepth },
    );

    return result.data;
  }

  /**
   * Query provenance records backward to find origins.
   *
   * Traces the lineage of data backward in time to discover
   * its original sources and the chain of transformations.
   *
   * @param sourceId - Source identifier to trace backward from
   * @param maxDepth - Maximum depth of backward traversal (default: 10)
   * @returns Backward provenance chain to origins
   */
  async queryBackward(
    sourceId: string,
    maxDepth: number = 10,
  ): Promise<ProvenanceQueryResult> {
    this.validateSourceId(sourceId);
    this.validateDepth(maxDepth);

    const result = await this.client.callTool<ProvenanceQueryResult>(
      SERVER,
      'provenance_query_backward',
      { source_id: sourceId, max_depth: maxDepth },
    );

    return result.data;
  }

  /**
   * Verify the provenance chain integrity for a source.
   *
   * Checks that the complete provenance chain for a data source
   * is consistent, with valid actor roles, timestamps, and
   * hash chains for transformed data.
   *
   * @param sourceId - Source identifier to verify
   * @returns Verification result with chain details
   */
  async verify(sourceId: string): Promise<ProvenanceVerifyResult> {
    this.validateSourceId(sourceId);

    const result = await this.client.callTool<ProvenanceVerifyResult>(
      SERVER,
      'provenance_verify',
      { source_id: sourceId },
    );

    return result.data;
  }

  /**
   * Get a complete lineage report for a source.
   * Combines forward and backward queries with verification.
   */
  async fullLineage(sourceId: string): Promise<{
    forward: ProvenanceQueryResult;
    backward: ProvenanceQueryResult;
    verification: ProvenanceVerifyResult;
  }> {
    const [forward, backward, verification] = await Promise.all([
      this.queryForward(sourceId),
      this.queryBackward(sourceId),
      this.verify(sourceId),
    ]);

    return { forward, backward, verification };
  }

  /**
   * Get all unique actors who have accessed a data source.
   */
  async getActors(sourceId: string): Promise<string[]> {
    const verification = await this.verify(sourceId);
    return verification.actors_involved;
  }

  // -------------------------------------------------------------------------
  // Validation
  // -------------------------------------------------------------------------

  private validateRecordParams(params: ProvenanceRecordParams): void {
    if (!params.source_id || params.source_id.trim().length === 0) {
      throw new InvalidInputError('source_id is required', {
        server: SERVER,
        tool: 'provenance_record_access',
        details: { field: 'source_id' },
      });
    }

    if (!params.action || params.action.trim().length === 0) {
      throw new InvalidInputError('action is required', {
        server: SERVER,
        tool: 'provenance_record_access',
        details: { field: 'action' },
      });
    }

    if (!params.actor_id || params.actor_id.trim().length === 0) {
      throw new InvalidInputError('actor_id is required', {
        server: SERVER,
        tool: 'provenance_record_access',
        details: { field: 'actor_id' },
      });
    }

    if (!VALID_ROLES.includes(params.actor_role)) {
      throw new InvalidInputError(
        `Invalid actor_role '${params.actor_role}'. Must be one of: ${VALID_ROLES.join(', ')}`,
        {
          server: SERVER,
          tool: 'provenance_record_access',
          details: { field: 'actor_role' },
        },
      );
    }

    if (!params.tool_call || params.tool_call.trim().length === 0) {
      throw new InvalidInputError('tool_call is required', {
        server: SERVER,
        tool: 'provenance_record_access',
        details: { field: 'tool_call' },
      });
    }
  }

  private validateSourceId(sourceId: string): void {
    if (!sourceId || sourceId.trim().length === 0) {
      throw new InvalidInputError('source_id is required', {
        server: SERVER,
        details: { field: 'source_id' },
      });
    }
  }

  private validateDepth(maxDepth: number): void {
    if (maxDepth < 1 || maxDepth > 100) {
      throw new InvalidInputError(
        `max_depth must be between 1 and 100, got ${maxDepth}`,
        { server: SERVER, details: { field: 'max_depth' } },
      );
    }
  }
}
